### ISSUE TITLE
rpmbuild output is garbled

### ISSUE THREAD

**Issue by voxik** [rpm-software-management/rpm#794]
Recently, presumably since RPM 4.15.0.beta.1 landed in Fedora Rawhide, the rpmbuild output is garbled. E.g.:

~~~
... snip ...
1)
Process.clock_getres matches the clock in practice for Process::CLOCK_PROCESS_CPUTIME_ID FAILED
Expected 10000
 not to equal 10000
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:30:in `block (4 levels) in <top (required)>'
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:4:in `<top (required)>'
2)
BUILDSTDERR: Process.clock_getres matches the clock in practicmake: *** [uncommon.mk:813: yes-test-spec] Error 1
BUILDSTDERR: error: Bad exit status from /var/tmp/rpm-tmp.WeWwGq (%check)
e for Process::CLOCK_THREAD_CPUTIME_ID FAILED
Expected 10000
 not to equal 10000
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:30:in `block (4 levels) in <top (required)>'
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:4:in `<top (required)>'
Finished in 189.216527 seconds
3770 files, 30971 examples, 112151 expectations, 2 failures, 0 errors, 0 tagged
RPM build errors:
BUILDSTDERR:     Bad exit status from /var/tmp/rpm-tmp.WeWwGq (%check)
Child return code was: 1
EXCEPTION: [Error()]
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/mockbuild/trace_decorator.py", line 96, in trace
    result = func(*args, **kw)
  File "/usr/lib/python3.7/site-packages/mockbuild/util.py", line 736, in do_with_status
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
 # bash --login -c /usr/bin/rpmbuild -bb --target armv7hl --nodeps /builddir/build/SPECS/ruby.spec
~~~

The correct output should IMO be:

~~~
1)
Process.clock_getres matches the clock in practice for Process::CLOCK_PROCESS_CPUTIME_ID FAILED
Expected 10000
 not to equal 10000
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:30:in `block (4 levels) in <top (required)>'
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:4:in `<top (required)>'
2)
Process.clock_getres matches the clock in practice for Process::CLOCK_THREAD_CPUTIME_ID FAILED
Expected 10000
 not to equal 10000
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:30:in `block (4 levels) in <top (required)>'
/builddir/build/BUILD/ruby-2.7.0-0c6c937904/spec/ruby/core/process/clock_getres_spec.rb:4:in `<top (required)>'
Finished in 189.216527 seconds
3770 files, 30971 examples, 112151 expectations, 2 failures, 0 errors, 0 tagged
RPM build errors:
BUILDSTDERR: make: *** [uncommon.mk:813: yes-test-spec] Error 1
BUILDSTDERR: error: Bad exit status from /var/tmp/rpm-tmp.WeWwGq (%check)
BUILDSTDERR:     Bad exit status from /var/tmp/rpm-tmp.WeWwGq (%check)
Child return code was: 1
EXCEPTION: [Error()]
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/mockbuild/trace_decorator.py", line 96, in trace
    result = func(*args, **kw)
  File "/usr/lib/python3.7/site-packages/mockbuild/util.py", line 736, in do_with_status
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
 # bash --login -c /usr/bin/rpmbuild -bb --target armv7hl --nodeps /builddir/build/SPECS/ruby.spec
~~~

It might also be issue of mock ....

**Comment by ignatenkobrain** [comment]
I think this is due to changes in mock (don't know which ones though). I don't see such behavior locally..

**Comment by ignatenkobrain** [comment]
Just to clarify, I don't see this behavior with plain rpmbuild

**Comment by voxik** [comment]
@xsuchy any idea?

**Comment by voxik** [comment]
Actually there was thread about this on fedora-devel:

https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/6LHINQVZYABXBKAEN3BQ25JPMB5FYEI2/

CC @hroncok

**Comment by hroncok** [comment]
Another example:

```
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.Ds3VBv
+ umask 022
+ cd /builddir/build/BUILD
+ cd pip-19.1.1
+ pytest_k='not completion and
BUILDSTDERR:           not test_pep517_and_build_options and
BUILDSTDERR:           not test_config_file_venv_option and
BUILDSTDERR:           not test_uninstall_non_local_distutils'
+ mkdir _bin
+ export PATH=/builddir/build/BUILD/pip-19.1.1/_bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin
+ PATH=/builddir/build/BUILD/pip-19.1.1/_bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin
+ export PYTHONPATH=/builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python2.7/site-packages
+ PYTHONPATH=/builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python2.7/site-packages
+ ln -s /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/bin/pip2 _bin/pip
BUILDSTDERR: ++ echo not completion and not test_pep517_and_build_options and not test_config_file_venv_option and not test_uninstall_non_local_distutils
+ /usr/bin/python2 -m pytest -m 'not network' -k 'not completion and not test_pep517_and_build_options and not test_config_file_venv_option and not test_uninstall_non_local_distutils'
============================= test session starts ==============================
platform linux2 -- Python 2.7.16, pytest-4.4.1, py-1.8.0, pluggy-0.11.0
rootdir: /builddir/build/BUILD/pip-19.1.1, inifile: setup.cfg
collected 1177 items / 125 deselected / 1052 selected
tests/functional/test_broken_stdout.py ...                               [  0%]
tests/functional/test_check.py ..........                                [  1%]
tests/functional/test_configuration.py .s..                              [  1%]
tests/functional/test_download.py .......................                [  3%]
tests/functional/test_freeze.py ......x...s......                        [  5%]
tests/functional/test_hash.py ...                                        [  5%]
tests/functional/test_help.py .......                                    [  6%]
tests/functional/test_install.py .........................ss..........x. [ 10%]
..............................ss............                             [ 14%]
tests/functional/test_install_check.py ..                                [ 14%]
tests/functional/test_install_cleanup.py .....                           [ 14%]
tests/functional/test_install_compat.py .                                [ 15%]
tests/functional/test_install_config.py .....                            [ 15%]
tests/functional/test_install_extras.py .....                            [ 15%]
tests/functional/test_install_force_reinstall.py ..                      [ 16%]
tests/functional/test_install_index.py ....                              [ 16%]
tests/functional/test_install_reqs.py .........................x...      [ 19%]
tests/functional/test_install_upgrade.py ........                        [ 20%]
tests/functional/test_install_user.py ..                                 [ 20%]
tests/functional/test_install_vcs_git.py ...............                 [ 21%]
tests/functional/test_install_wheel.py .........................         [ 24%]
tests/functional/test_list.py ............                               [ 25%]
tests/functional/test_no_color.py .                                      [ 25%]
tests/functional/test_pep517.py ..........                               [ 26%]
tests/functional/test_requests.py s                                      [ 26%]
tests/functional/test_search.py ....                                     [ 26%]
tests/functional/test_show.py ...............                            [ 28%]
tests/functional/test_uninstall.py ...........                           [ 29%]
tests/functional/test_uninstall_user.py ..                               [ 29%]
tests/functional/test_vcs_bazaar.py s..                                  [ 29%]
tests/functional/test_vcs_git.py ........                                [ 30%]
tests/functional/test_warning.py .                                       [ 30%]
tests/functional/test_wheel.py ................                          [ 32%]
tests/functional/test_yaml.py x......xx.....                             [ 33%]
tests/lib/test_lib.py .................                                  [ 34%]
tests/unit/test_appdirs.py ......................                        [ 37%]
tests/unit/test_base_command.py ......                                   [ 37%]
tests/unit/test_build_env.py .....                                       [ 38%]
tests/unit/test_cache.py ..                                              [ 38%]
tests/unit/test_check.py ..                                              [ 38%]
tests/unit/test_command_install.py ..                                    [ 38%]
tests/unit/test_compat.py ...........                                    [ 39%]
tests/unit/test_configuration.py ......................                  [ 41%]
tests/unit/test_download.py .........s........s..............            [ 44%]
tests/unit/test_finder.py .............................                  [ 47%]
tests/unit/test_format_control.py .........                              [ 48%]
tests/unit/test_index.py ............................................... [ 53%]
BUILDSTDERR: .........................s.                                       + export PYTHONPATH=/builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python3.7/site-packages
+ PYTHONPATH=/builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python3.7/site-packages
+ ln -sf /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/bin/pip3 _bin/pip
BUILDSTDERR: ++ echo not completion and not test_pep517_and_build_options and not test_config_file_venv_option and not test_uninstall_non_local_distutils
+ /usr/bin/python3 -m pytest -m 'not network' -k 'not completion and not test_pep517_and_build_options and not test_config_file_venv_option and not test_uninstall_non_local_distutils'
       [ 55%]
tests/unit/test_index_html_page.py .............                         [ 56%]
tests/unit/test_locations.py ...                                         [ 57%]
tests/unit/test_logging.py .............                                 [ 58%]
tests/unit/test_models.py .....                                          [ 58%]
tests/unit/test_options.py ............................................. [ 63%]
.........                                                                [ 63%]
tests/unit/test_pep425tags.py ..................                         [ 65%]
tests/unit/test_pep517.py .....                                          [ 66%]
tests/unit/test_req.py ........................................          [ 69%]
tests/unit/test_req_file.py ............................................ [ 74%]
................                                                         [ 75%]
tests/unit/test_req_install.py .....                                     [ 76%]
tests/unit/test_req_uninstall.py ..........                              [ 77%]
tests/unit/test_unit_outdated.py ......                                  [ 77%]
tests/unit/test_utils.py ............................................... [ 82%]
...........................................................              [ 87%]
tests/unit/test_vcs.py s................................................ [ 92%]
..................                                                       [ 94%]
tests/unit/test_wheel.py ............................................... [ 98%]
...............                                                          [100%]
=============================== warnings summary ===============================
tests/unit/test_pep425tags.py::TestPEP425Tags::test_broken_sysconfig
  /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python2.7/site-packages/pip/_internal/pep425tags.py:34: RuntimeWarning: I have the wrong path!
    warnings.warn("{}".format(e), RuntimeWarning)
tests/unit/test_req_file.py::TestParseRequirements::test_install_requirements_with_options
  /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python2.7/site-packages/pip/_internal/req/req_file.py:184: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
    cmdoptions.check_install_build_global(options, opts)
-- Docs: https://docs.pytest.org/en/latest/warnings.html
=========================== short test summary info ============================
SKIPPED [1] tests/unit/test_vcs.py:23: Subversion is only required under Travis
SKIPPED [1] /builddir/build/BUILD/pip-19.1.1/tests/functional/test_install.py:1284: No lzma support
SKIPPED [1] tests/unit/test_download.py:174: condition: sys.platform != 'win32'
SKIPPED [1] tests/functional/test_install.py:331: Mercurial is not available
SKIPPED [1] tests/unit/test_download.py:206: condition: sys.platform != 'win32'
SKIPPED [1] tests/functional/test_install.py:322: Mercurial is not available
SKIPPED [1] tests/functional/test_requests.py:4: <Skipped instance>
SKIPPED [1] /builddir/build/BUILD/pip-19.1.1/tests/functional/test_install.py:1275: No lzma support
SKIPPED [1] tests/functional/test_configuration.py:19: Can't modify underlying file for any mode
SKIPPED [1] tests/functional/test_vcs_bazaar.py:15: Bazaar is only required under Travis
SKIPPED [1] tests/functional/test_freeze.py:367: Mercurial is not available
SKIPPED [1] tests/unit/test_index.py:321: condition: sys.platform != 'win32'
XFAIL tests/functional/test_freeze.py::test_freeze_exclude_editable
XFAIL tests/functional/test_install.py::test_install_argparse_shadowed
XFAIL tests/functional/test_install_reqs.py::test_install_distribution_union_conflicting_extras
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/conflicting_diamond]
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/extras-2]
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/conflicting_triangle]
= 1034 passed, 12 skipped, 125 deselected, 6 xfailed, 2 warnings in 1555.48 seconds =
============================= test session starts ==============================
platform linux -- Python 3.7.4, pytest-4.4.1, py-1.8.0, pluggy-0.12.0
rootdir: /builddir/build/BUILD/pip-19.1.1, inifile: setup.cfg
collected 1177 items / 125 deselected / 1052 selected
tests/functional/test_broken_stdout.py ...                               [  0%]
tests/functional/test_check.py ..........                                [  1%]
tests/functional/test_configuration.py .s..                              [  1%]
tests/functional/test_download.py .......................                [  3%]
tests/functional/test_freeze.py ......x...s......                        [  5%]
tests/functional/test_hash.py ...                                        [  5%]
tests/functional/test_help.py .......                                    [  6%]
tests/functional/test_install.py .........................ss..........s. [ 10%]
............................................                             [ 14%]
tests/functional/test_install_check.py ..                                [ 14%]
tests/functional/test_install_cleanup.py .....                           [ 14%]
tests/functional/test_install_compat.py .                                [ 15%]
tests/functional/test_install_config.py .....                            [ 15%]
tests/functional/test_install_extras.py .....                            [ 15%]
tests/functional/test_install_force_reinstall.py ..                      [ 16%]
tests/functional/test_install_index.py ....                              [ 16%]
tests/functional/test_install_reqs.py .........................x...      [ 19%]
tests/functional/test_install_upgrade.py ........                        [ 20%]
tests/functional/test_install_user.py ..                                 [ 20%]
tests/functional/test_install_vcs_git.py ...............                 [ 21%]
tests/functional/test_install_wheel.py .........................         [ 24%]
tests/functional/test_list.py ............                               [ 25%]
tests/functional/test_no_color.py .                                      [ 25%]
tests/functional/test_pep517.py ..........                               [ 26%]
tests/functional/test_requests.py s                                      [ 26%]
tests/functional/test_search.py ....                                     [ 26%]
tests/functional/test_show.py ...............                            [ 28%]
tests/functional/test_uninstall.py ...........                           [ 29%]
tests/functional/test_uninstall_user.py ..                               [ 29%]
tests/functional/test_vcs_bazaar.py s..                                  [ 29%]
tests/functional/test_vcs_git.py ........                                [ 30%]
tests/functional/test_warning.py .                                       [ 30%]
tests/functional/test_wheel.py ................                          [ 32%]
tests/functional/test_yaml.py x......xx.....                             [ 33%]
tests/lib/test_lib.py .................                                  [ 34%]
tests/unit/test_appdirs.py ......................                        [ 37%]
tests/unit/test_base_command.py ......                                   [ 37%]
tests/unit/test_build_env.py .....                                       [ 38%]
tests/unit/test_cache.py ..                                              [ 38%]
tests/unit/test_check.py ..                                              [ 38%]
tests/unit/test_command_install.py ..                                    [ 38%]
tests/unit/test_compat.py ...........                                    [ 39%]
tests/unit/test_configuration.py ......................                  [ 41%]
tests/unit/test_download.py .........s........s..............            [ 44%]
tests/unit/test_finder.py .............................                  [ 47%]
tests/unit/test_format_control.py .........                              [ 48%]
tests/unit/test_index.py ............................................... [ 53%]
.........................s.                                              [ 55%]
BUILDSTDERR: tests/unit/test_in+ RPM_EC=0
BUILDSTDERR: ++ jobs -p
+ exit 0
dex_html_page.py .............                         [ 56%]
tests/unit/test_locations.py ...                                         [ 57%]
tests/unit/test_logging.py .............                                 [ 58%]
tests/unit/test_models.py .....                                          [ 58%]
tests/unit/test_options.py ............................................. [ 63%]
.........                                                                [ 63%]
tests/unit/test_pep425tags.py ..................                         [ 65%]
tests/unit/test_pep517.py .....                                          [ 66%]
tests/unit/test_req.py ........................................          [ 69%]
tests/unit/test_req_file.py ............................................ [ 74%]
................                                                         [ 75%]
tests/unit/test_req_install.py .....                                     [ 76%]
tests/unit/test_req_uninstall.py ..........                              [ 77%]
tests/unit/test_unit_outdated.py ......                                  [ 77%]
tests/unit/test_utils.py ............................................... [ 82%]
...........................................................              [ 87%]
tests/unit/test_vcs.py s................................................ [ 92%]
..................                                                       [ 94%]
tests/unit/test_wheel.py ............................................... [ 98%]
...............                                                          [100%]
=============================== warnings summary ===============================
tests/unit/test_pep425tags.py::TestPEP425Tags::test_broken_sysconfig
  /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python3.7/site-packages/pip/_internal/pep425tags.py:34: RuntimeWarning: I have the wrong path!
    warnings.warn("{}".format(e), RuntimeWarning)
tests/unit/test_req_file.py::TestParseRequirements::test_install_requirements_with_options
  /builddir/build/BUILDROOT/python-pip-19.1.1-3.fc31.noarch/usr/lib/python3.7/site-packages/pip/_internal/req/req_file.py:184: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
    cmdoptions.check_install_build_global(options, opts)
-- Docs: https://docs.pytest.org/en/latest/warnings.html
=========================== short test summary info ============================
SKIPPED [1] tests/functional/test_configuration.py:19: Can't modify underlying file for any mode
SKIPPED [1] tests/functional/test_freeze.py:367: Mercurial is not available
SKIPPED [1] tests/functional/test_install.py:322: Mercurial is not available
SKIPPED [1] tests/functional/test_install.py:331: Mercurial is not available
SKIPPED [1] tests/functional/test_install.py:531: condition: sys.version_info >= (3,4)
SKIPPED [1] tests/functional/test_requests.py:4: <Skipped instance>
SKIPPED [1] tests/functional/test_vcs_bazaar.py:15: Bazaar is only required under Travis
SKIPPED [1] tests/unit/test_download.py:174: condition: sys.platform != 'win32'
SKIPPED [1] tests/unit/test_download.py:206: condition: sys.platform != 'win32'
SKIPPED [1] tests/unit/test_index.py:321: condition: sys.platform != 'win32'
SKIPPED [1] tests/unit/test_vcs.py:23: Subversion is only required under Travis
XFAIL tests/functional/test_freeze.py::test_freeze_exclude_editable
XFAIL tests/functional/test_install_reqs.py::test_install_distribution_union_conflicting_extras
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/conflicting_diamond]
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/extras-2]
XFAIL tests/functional/test_yaml.py::test_yaml_based[install/conflicting_triangle]
= 1036 passed, 11 skipped, 125 deselected, 5 xfailed, 2 warnings in 2444.67 seconds =
```

**Comment by hroncok** [comment]
Disclaimer: I never build in local rpmbuild, so I have no idea if this is caused by mock or RPM.

**Comment by xsuchy** [comment]
I am not aware of any related change in Mock in the past ~1 year. The relevant Mock code is here:
https://github.com/rpm-software-management/mock/blob/devel/mock/py/mockbuild/util.py#L555

Hmmm, what changed is that mock now runs in Python3 - even on EPEL7. I am going to check if `read()` changed behavior in some way.

**Comment by hroncok** [comment]
I observed the behavior change on Fedora rawhide only.

**Comment by xsuchy** [comment]
Can you share src.rpm which produce the error above? I have either too simple specs or without any error.

**Comment by hroncok** [comment]
This was python-pip (use the Fedora rawhide one).

**Comment by xsuchy** [comment]
I used that python-pip from dist-git and with mock -r fedora-rawhide-x86_64 I was able to reproduce. Then I `mock --copyin python-pip.src.rpm` and `mock -r fedora-rawhide-x86_64 shell` and `cd /buildir/build/SRPMS; rpmbuild --rebuild python-pip.src.rpm` and I do not see this issue. So it really seems that mock is in fault.

**Comment by voxik** [comment]
Well, I'd blame the RPM multithreading modifications, that is why I have reported it here.

**Comment by ffesti** [comment]
Well, the multi threaded build should only affect the creating of the binary packages (including srpm) but not the build scripts which are - of course -  still executed in order.
I suspect the changes in IO handling done for the dynamic build requires as they change how stdout is handled for build scripts. Any change this could be caused by mixing stdout and stderr in a different way?

**Comment by pmatilai** [comment]
Yeah we probably should stop grabbing stdout on our own when we dont actually use the data. Besides causing various display-artifacts under mock and the like, it's also simply doing a whole lot of work for no reason.

**Comment by pmatilai** [comment]
This should be fixed by PR #818. If you can test, that'd be awesome.
